gemma/gemma-7b-it-trtllm/packages/client.py

```diff
@@ -0,0 +1,155 @@
+import json
+import os
+import subprocess
+import time
+from functools import partial
+from pathlib import Path
+from queue import Queue
+from threading import Thread
+
+import tritonclient.grpc as grpcclient
+import tritonclient.http as httpclient
+from tritonclient.utils import InferenceServerException
+from utils import (
+    GRPC_SERVICE_PORT,
+    HTTP_SERVICE_PORT,
+    prepare_model_repository,
+    server_loaded,
+)
+
+
+class UserData:
+    def __init__(self):
+        self._completed_requests = Queue()
+
+
+def callback(user_data, result, error):
+    if error:
+        user_data._completed_requests.put(error)
+    else:
+        user_data._completed_requests.put(result)
+
+
+class TritonClient:
+    def __init__(self, data_dir: Path, model_repository_dir: Path, parallel_count=1):
+        self._data_dir = data_dir
+        self._model_repository_dir = model_repository_dir
+        self._parallel_count = parallel_count
+        self._http_client = None
+        self._grpc_client_map = {}
+
+    def start_grpc_stream(self, user_data, model_name, inputs, stream_uuid):
+        """Starts a GRPC stream and sends a request to the Triton server."""
+        grpc_client_instance = grpcclient.InferenceServerClient(
+            url=f"localhost:{GRPC_SERVICE_PORT}", verbose=False
+        )
+        self._grpc_client_map[stream_uuid] = grpc_client_instance
+        grpc_client_instance.start_stream(callback=partial(callback, user_data))
+        grpc_client_instance.async_stream_infer(
+            model_name,
+            inputs,
+            request_id=stream_uuid,
+            enable_empty_final_response=True,
+        )
+
+    def stop_grpc_stream(self, stream_uuid, stream_thread: Thread):
+        """Closes a GRPC stream and stops the associated thread."""
+        triton_grpc_stream = self._grpc_client_map[stream_uuid]
+        triton_grpc_stream.stop_stream()
+        stream_thread.join()
+        del self._grpc_client_map[stream_uuid]
+
+    def start_server(
+        self,
+        mpi: int = 1,
+        env: dict = {},
+    ):
+        """Triton Inference Server has different startup commands depending on
+        whether it is running in a TP=1 or TP>1 configuration. This function
+        starts the server with the appropriate command."""
+        if mpi == 1:
+            command = [
```

Comment:
instead of manually building up this list of strings iteratively, a more readable/scalable approach here could be adding a level of indirection, and instead builiding up some pydantic/dataclass object. Then you could have a function that converts that object into the bash command to run.

that way this TritonClient code does not have to concern itself with formatting strings correctly/getting argument order correct.